src/main/scala/regmapper/Annotation.scala

```diff
@@ -2,17 +2,121 @@
 
 package freechips.rocketchip.regmapper
 
+import chisel3.internal.InstanceId
+import chisel3.experimental.{ChiselAnnotation, RunFirrtlTransform, annotate}
+import firrtl.{CircuitForm, CircuitState, LowForm, Transform}
+import firrtl.annotations._
+import freechips.rocketchip.regmapper.RegFieldAccessType.RegFieldAccessType
+import freechips.rocketchip.regmapper.RegFieldRdAction.RegFieldRdAction
+import freechips.rocketchip.regmapper.RegFieldWrType.RegFieldWrType
+import org.json4s.JsonAST.JValue
 import org.json4s.JsonDSL._
 import org.json4s.jackson.JsonMethods.{pretty, render}
 
+// Do not call this function
+// This function is called by DescribedRegChiselAnnotation
+case class RegFieldDescAnnotation(
+    target: Named,
+    desc: String) extends SingleTargetAnnotation[Named] {
+  def duplicate(n: Named): RegFieldDescAnnotation = this.copy(target = n)
+}
+
+object RegFieldDescAnnotation {
+  def nameMap(state: CircuitState): Map[String,String] = {
+    state.annotations.collect {
+      case RegFieldDescAnnotation(mname, cname) => mname.toString -> cname
+    }.toMap
+  }
+}
+
+
+case class DescribedRegChiselAnnotation(
+  target: InstanceId,
+  json: String) extends ChiselAnnotation with RunFirrtlTransform {
+  def toFirrtl: RegFieldDescAnnotation = {
+
+    RegFieldDescAnnotation(target.toNamed, json)
+  }
+
+  override def transformClass: Class[_ <: Transform] = classOf[DescribedRegDumpTransform]
+}
+
+object DescribedRegChiselAnnotationUtil {
+  def toJson (
+//    target: InstanceId,
+    regField: freechips.rocketchip.regmapper.RegField,
+    byteOffset: Int,
+    bitOffset: Int): JValue = {
+
+    val regDesc: freechips.rocketchip.regmapper.RegFieldDesc = regField.desc.get
+
+    //require("", regDesc.desc != None)
+
+    ( ("byteOffset"   -> s"0x${byteOffset.toHexString}") ~
+      ("bitOffset"    -> bitOffset) ~
+      ("bitWidth"     -> regField.width) ~
+      ("name"         -> regDesc.name) ~
+      ("description"  -> regDesc.desc) ~
+// TODO      ("resetValue"   -> regDesc.reset.getOrElse("0")) ~
```

Comment:
why can't this handle optional parameters?

Reply:
Optional params are added but I am not sure what I should put for the default param values in the Else.

Comments?

Reply:
I don't understand why you need anythng if it's optional. If it's not there, then there IS NO reset value. I don't know what you are doing with this so what the limitation is. If this is emitting JSON as it was before, then if the resetValue is omitted from the JSON, then the field is not reset.